tests/foreman/api/test_hostgroup.py

```diff
@@ -148,6 +148,64 @@ def test_bz_1107708(self):
         self.assertEqual(host_attrs['all_puppetclasses'][0]['name'], 'ntp')
 
 
+@skip_if_bug_open('bugzilla', 1235377)
+class MissingAttrTestCase(APITestCase):
+    """Tests to see if the server returns the attributes it should."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Create a ``HostGroup`` as ``cls.host_group``."""
+        cls.host_group = entities.HostGroup().create()
+
+    def test_get_content_source(self):
+        """@Test: Read a host group. Inspect the server's response.
+
+        @Assert: The response contains either a ``content_source`` or
+        ``content_source_id`` attribute.
+
+        @Feature: HostGroup
+
+        """
+        wanted_attrs = ('content_source', 'content_source_id')
```

Comment:
There's two possible changes to make here, and the two changes can be made independently. The first change is to call `read_json` in `setUpClass`. What's the impact? Before the change, the test looks like this:

```python
    def test_get_content_source(self):
        wanted_attrs = ('content_source', 'content_source_id')
        attrs = self.host_group.read_json()
        self.assertTrue(
            any([wanted_attr in attrs for wanted_attr in wanted_attrs]),
            'None of {0} are in {1}'.format(wanted_attrs, attrs.keys())
        )
```

After the change, the test looks like this:

```python
    def test_get_content_source(self):
        attrs = ('content_source', 'content_source_id')
        self.assertTrue(
            any([attr in self.hg_attrs for attr in attrs]),
            'None of {0} are in {1}'.format(attrs, self.hg_attrs.keys())
        )
```

We've dropped a line of code from the test and reduced the test execution time by getting rid of those redundant `read_json` calls.

The second possible change is to use `_get_entity_id`. The `_get_entity_id` method does the following:

* It generates the permutations of a field name. For example, given 'content_source' it generates `content_source` and `content_source_id`.
* It checks for the presence of any one of those permutations in a dict of values.

With just the second change in place, the test reduces to:

```python
    def test_get_content_source(self):
        _get_entity_id('content_source', self.host_group.read_json())
```

With both the first and second changes in place, the test reduces to:

```python
    def test_get_content_source(self):
        _get_entity_id('content_source', self.host_group_attrs)
```

Does that explain things, @omaciel?